Get Citus compiling and passing on PostgreSQL 10

jasonmp85 opened the pull request:
There is still some room for improvement in these changes (in particular, I think a lot of the places we test `PG_VERSION_NUM` should be abstracted out and cleaned up in all C files), but everything compiles and passes.

PostgreSQL 10 changes the output of `psql`'s `\d`, but we use it in so many places (over 170), that I didn't want to create test variants for every file using it. @anarazel suggested rewriting all tests that use it, and I have an SQL alternative to describing a table, but there are still tests which require foreign key, constraint, or index outputs, which I have yet to polyfill.

So for now (and this is temporary), I've locked the `psql` version to `9.6` to get the old behavior (though you won't see that here, it's in the `citusdata/tools` branch this code uses to build on PostgreSQL 10).

jasonmp85 commented:
FWIW the change in coverage is due to not testing with the HLL extension present, near as I can tell.

jasonmp85 commented:
Ok, so for completeness, I built HLL locally for PostgreSQL 10 (needed one include line added) and installed it. The Citus test that uses it works great in PostgreSQL 10, so I don't anticipate any problems in the handling of those aggregates, etc.

anarazel commented on src/backend/distributed/planner/multi_router_planner.c:
```diff
@@ -1638,13 +1641,14 @@ MasterIrreducibleExpressionWalker(Node *expression, WalkerState *state)
 	 */
-	StaticAssertStmt(PG_VERSION_NUM < 90700, "When porting to a newer PG this section"
-											 " needs to be reviewed.");
+	StaticAssertStmt(PG_VERSION_NUM < 100100, "When porting to a newer PG this section"
```
Have you compared whether this actually needs additions?  From the back of my mind that doesn't sound unlikely.

anarazel replied:
Ping?

jasonmp85 replied:
I verified all of this when making the change.

anarazel commented on src/backend/distributed/utils/ruleutils_10.c:
```diff
@@ -0,0 +1,7745 @@
+ *
+ *
+ * IDENTIFICATION
+ *	  src/backend/distributed/utils/ruleutils_96.c
```
Wrong filename - how did you generate the file?

jasonmp85 replied:
This is a typo.

I copied the file from PostgreSQL 10 then applied the Citus changes (the commit history should reflect this).

I'll bet I just region-copied the whole Citus header and forgot to modify the version in the filename (so no worries that I accidentally e.g. copied 9.6 and called it 10).

anarazel commented on src/backend/distributed/worker/worker_data_fetch_protocol.c:
```diff
@@ -1149,6 +1154,22 @@ ExecuteRemoteCommand(const char *nodeName, uint32 nodePort, StringInfo queryStri
+ */
+Node *
+ParseTreeRawStmt(const char *ddlCommand)
+{
```
This code makes me very sad, everytime I look at it :(.  Obviously not this commit's fault.

aamederen replied:
Why?

anarazel replied:
Because using raw parsetrees is just about always wrong.  It's really hard to correctly work with them, because no semantic analysis has happened yet, and we can't do semantic analysis at that point, because referenced relations don't exist.  Because it happens inside of functions, we need separate infrastructure for CONCURRENTLY, VACUUM etc.

anarazel commented on src/backend/distributed/worker/worker_data_fetch_protocol.c:
```diff
@@ -417,8 +422,8 @@ worker_apply_shard_ddl_command(PG_FUNCTION_ARGS)
-	ProcessUtility(ddlCommandNode, ddlCommand, PROCESS_UTILITY_TOPLEVEL,
-				   NULL, None_Receiver, NULL);
+	CitusProcessUtility(ddlCommandNode, ddlCommand, PROCESS_UTILITY_TOPLEVEL, NULL,
+						None_Receiver, NULL);
```
Not this commit's fault, but why on earth is it correct to use PROCESS_UTILITY_TOPLEVEL here?  Isn't that what makes it easy to crash citus by sending a CONCURRENTLY command w/ worker_apply_shard_ddl_command?

anarazel commented on src/backend/distributed/worker/worker_merge_protocol.c:
```diff
@@ -340,7 +339,16 @@ RemoveJobSchema(StringInfo schemaName)
+						PERFORM_DELETION_QUIETLY |
+						PERFORM_DELETION_SKIP_ORIGINAL |
+						PERFORM_DELETION_SKIP_EXTENSIONS);
+#else
```
Couldn't we just always use performDeletion?

jasonmp85 replied:
No, these flags were added in postgres/postgres@b3427dade14cc31eb48740bc9ea98b5954470b24 so emulating the behavior of `deleteWhatDependsUpon` is not possible in prior PostgreSQL versions.

anarazel commented on .travis.yml:
```diff
@@ -11,16 +11,15 @@ env:
+    - PGVERSION=10
 before_install:
-  - git clone -b v0.5.1 --depth 1 https://github.com/citusdata/tools.git
+  - git clone -b pg10_hack --depth 1 https://github.com/citusdata/tools.git
```
I'm very staunchly opposed to merge this with this hack present.

jasonmp85 replied:
Branch still named that, hack removed.

jasonmp85 replied:
Done.

anarazel commented on src/backend/distributed/master/master_metadata_utility.c:
```diff
@@ -1211,9 +1233,13 @@ UpdateColocationGroupReplicationFactor(uint32 colocationId, int replicationFacto
+	CatalogTupleUpdate(pgDistColocation, &newHeapTuple->t_self, newHeapTuple);
+#else
+	simple_heap_update(pgDistColocation, &newHeapTuple->t_self, newHeapTuple);
 	CatalogUpdateIndexes(pgDistColocation, newHeapTuple);
```
PingPing.

anarazel replied:
PingPongPingPing.

jasonmp85 replied:
Sure, I'll write some sort of polyfill/compat shim.

jasonmp85 replied:
Done.

anarazel commented on .travis.yml:
```diff
@@ -13,18 +13,15 @@ env:
+    - PGVERSION=10
 before_install:
-  - git clone -b v0.6.1 --depth 1 https://github.com/citusdata/tools.git
+  - git clone -b pg10_hack --depth 1 https://github.com/citusdata/tools.git
```
So this needs to be properly merged first?   What's the schedule there? Any issues?

jasonmp85 replied:
I opened citusdata/tools#17 to show what's changed.

Basically if those changes look good, I'll merge and release `0.7` or something and set Citus to use that in `master` (as `9.5` and `9.6` behavior is unaffected).

jasonmp85 replied:
Done.

anarazel commented on .travis.yml:
```diff
@@ -13,18 +13,15 @@ env:
+  - git clone -b pg10_hack --depth 1 https://github.com/citusdata/tools.git
   - sudo make -C tools install
   - setup_apt
-  - curl https://install.citusdata.com/community/deb.sh | sudo bash
```
Hm?  There's community debs for 10 already, so what's the deal here?

jasonmp85 replied:
What do you mean? I don't think we've released any 10-compatible extensions or packages thus far… Should I build some against `10beta1` or something?

anarazel replied:
Doesn't this also setup pgdg for $version?

jasonmp85 replied:
No, it only sets up `pgdg` for `9.6`, and only if it's not already present, as it's meant to be the easiest user experience for new Citus users. If users need `9.5` or something, they need to set that up themselves.

`setup_apt` is the script which has always set up our Travis box to know about PGDG; it predates `install.citusdata.com` by quite a bit. It's been augmented to add the `10 main` repo.

jasonmp85 replied:
I don't believe there were any changes required from this thread.

anarazel commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -164,6 +164,40 @@ static void PostProcessUtility(Node *parsetree);
+	plannedStmt->commandType = CMD_UTILITY;
+	plannedStmt->utilityStmt = parsetree;
+
+	multi_ProcessUtility10(plannedStmt, queryString, context, params, NULL, dest,
```
Ping.

jasonmp85 replied:
Will rework and push up something nicer.

jasonmp85 replied:
Renamed the old function to 9x to reflect its use by the 9.x-line of PostgreSQL releases. The logic now simply lives in `multi_processUtility`.

anarazel commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -280,10 +326,18 @@ multi_ProcessUtility(Node *parsetree,
+			 * copy parse tree since we might scribble on it to fix the schema name
+			 * FIXME: PostgreSQL 10 changes caching behavior and breaks this approach,
+			 *        so for that version we do assign back the modified tree.
+			 */
```
So this surely isn't something we can merge?

jasonmp85 replied:
See the commit message for b2d5e3c and Marco's email about the topic.

At any rate, I just noticed this comment has a typo.

anarazel replied:
@marcocitus @jasonmp85 I'm not really ok with just regressing here.  The likelihood of tackling that anytime soon seems close to zero, and it'll get copied around making the problem harder and harder to fix.

jasonmp85 replied:
Rewrote this commit to copy/create strings in the same context as the struct we receive. This fixes the problem.

anarazel commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -2142,8 +2203,15 @@ ErrorIfDistributedAlterSeqOwnedBy(AlterSeqStmt *alterSeqStmt)
-	/* see whether the sequences is already owned by a distributed table */
-	if (sequenceIsOwned(sequenceId, &ownedByTableId, &ownedByColumnId))
+#if (PG_VERSION_NUM >= 100000)
+	sequenceOwned = sequenceIsOwned(sequenceId, DEPENDENCY_AUTO, &ownedByTableId,
```
I think it'd be better if we also checked for INTERNAL sequences (used by identity columns) here, for some future proofing.  For that matter, have you done any checking how identity columns work with citus and/or other v10 features?

anarazel replied:
Ping? Afaics this hasn't yet been addressed?  At the least it seems we need a test checking that identity type columns are handled without unexpected errors (i.e. by erroring out while distributing such tables instead).

jasonmp85 replied:
Gah. I was bound to miss something. Yes, I missed this; will come back with what you've asked for.

anarazel commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -2974,8 +3043,12 @@ PostProcessUtility(Node *parsetree)
+#else
 	simple_heap_update(pg_index, &indexTuple->t_self, indexTuple);
 	CatalogUpdateIndexes(pg_index, indexTuple);
+#endif
```
I still don't think it's good to do this, and that we should just add our own CatalogTupleUpdate() etc wrappers for the older releases.  I'm pretty sure I pointed this out before.

jasonmp85 replied:
Done.

anarazel commented on src/backend/distributed/shared_library_init.c:
```diff
@@ -154,7 +154,11 @@ _PG_init(void)
 
 	/* register utility hook */
+#if (PG_VERSION_NUM >= 100000)
+	ProcessUtility_hook = multi_ProcessUtility10;
```
Having 10 in the name will be a bit awkward for 11+, so maybe name the wrapper Pre10 or such?

jasonmp85 replied:
Named it `multi_ProcessUtility9x` because it looks cool.

![c60](https://user-images.githubusercontent.com/50673/27050480-ed81de1e-4f6f-11e7-9f00-ac6b0a118801.png)

onderkalaci commented:
I've been experimenting this branch. I don't know if it is important/critical or not, but wanted to note an issue.

```SQL
CREATE TABLE plist1(c1 NUMERIC, c2 VARCHAR(10)) PARTITION BY LIST (c1) ;

-- create two partition tables
CREATE TABLE plist1_v100 PARTITION OF plist1 FOR VALUES IN (100);
CREATE TABLE plist1_v200 PARTITION OF plist1 FOR VALUES IN (200); 

-- try to distribute partitioned table
-- awesome, Citus errored out correctly
SELECT create_distributed_table('plist1', 'c1');
ERROR:  cannot distribute relation: plist1
DETAIL:  Distributed relations must be regular or foreign tables.

-- try to distribute a partition table
-- Citus errors out, but, the message is not clear
SELECT create_distributed_table('plist1_v100', 'c1');
ERROR:  column "c1" of relation "plist1_v100_102032" does not exist
CONTEXT:  while executing command on localhost:9700
```

We're going to handle this in the scope of `shard by id` + `partition by time` ( #573 ), but, seems would be nice to error out on this PR when the user tries to distribute a partition table.

onderkalaci commented:
While experimenting this branch, another issue came up, which seems we'd probably want to fix before merging this.

Note that this only happens when Citus is in `shared_preload_libraries` and extension is created.

```SQL

CREATE TABLE prange1(id int, time date) PARTITION BY RANGE (time);

-- create its partitions
CREATE TABLE prange1_2006 PARTITION OF prange1 FOR VALUES FROM ('2006-01-01') TO ('2007-01-01') ;
CREATE TABLE prange1_2007 PARTITION OF prange1 FOR VALUES FROM ('2007-01-01') TO ('2008-01-01');

 CREATE INDEX CONCURRENTLY on prange1_2007(id);                                                                                                                                                                                                                       server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
The connection to the server was lost. Attempting reset: Failed.
!> 

```

Backtrace:
```
Program received signal EXC_BAD_ACCESS, Could not access memory.
Reason: KERN_INVALID_ADDRESS at address: 0x0000000000000000
0x00007fff9a0fd132 in strlen ()
(gdb) bt
#0  0x00007fff9a0fd132 in strlen ()
#1  0x00000001025fb98c in hashname (fcinfo=0x7fff5d65d7b0) at hashfunc.c:148
#2  0x0000000102c17669 in DirectFunctionCall1Coll (func=0x1025fb960 <hashname>, collation=0, arg1=0) at fmgr.c:717
#3  0x0000000102be6c98 in CatalogCacheComputeHashValue (cache=0x7fdd748368d0, nkeys=2, cur_skey=0x7fff5d65dcd0) at catcache.c:209
#4  0x0000000102be6143 in SearchCatCache (cache=0x7fdd748368d0, v1=0, v2=2200, v3=0, v4=0) at catcache.c:1130
#5  0x0000000102c01b94 in SearchSysCache (cacheId=45, key1=0, key2=2200, key3=0, key4=0) at syscache.c:1109
#6  0x0000000102c01ce3 in GetSysCacheOid (cacheId=45, key1=0, key2=2200, key3=0, key4=0) at syscache.c:1187
#7  0x0000000102c0547c in get_relname_relid (relname=0x0, relnamespace=2200) at lsyscache.c:1685
#8  0x00000001033556cd in PostProcessUtility (parsetree=0x7fdd74009e00) at executor/multi_utility.c:3063
#9  0x0000000103353e1a in multi_ProcessUtility10 (pstmt=0x7fdd7400a1c0, queryString=0x7fdd74009240 "CREATE INDEX CONCURRENTLY on prange1_2007_local(time);", context=PROCESS_UTILITY_TOPLEVEL, params=0x0, queryEnv=0x0, dest=0x7fdd7400a2b8, completionTag=0x7fff5d65e3d0 "") at executor/multi_utility.c:439
#10 0x0000000102a5b50c in ProcessUtility (pstmt=0x7fdd7400a1c0, queryString=0x7fdd74009240 "CREATE INDEX CONCURRENTLY on prange1_2007_local(time);", context=PROCESS_UTILITY_TOPLEVEL, params=0x0, queryEnv=0x0, dest=0x7fdd7400a2b8, completionTag=0x7fff5d65e3d0 "") at utility.c:353
#11 0x0000000102a5ae08 in PortalRunUtility (portal=0x7fdd73044440, pstmt=0x7fdd7400a1c0, isTopLevel=1 '\001', setHoldSnapshot=0 '\0', dest=0x7fdd7400a2b8, completionTag=0x7fff5d65e3d0 "") at pquery.c:1179
#12 0x0000000102a5a016 in PortalRunMulti (portal=0x7fdd73044440, isTopLevel=1 '\001', setHoldSnapshot=0 '\0', dest=0x7fdd7400a2b8, altdest=0x7fdd7400a2b8, completionTag=0x7fff5d65e3d0 "") at pquery.c:1325
#13 0x0000000102a59644 in PortalRun (portal=0x7fdd73044440, count=9223372036854775807, isTopLevel=1 '\001', run_once=1 '\001', dest=0x7fdd7400a2b8, altdest=0x7fdd7400a2b8, completionTag=0x7fff5d65e3d0 "") at pquery.c:800
#14 0x0000000102a54ecc in exec_simple_query (query_string=0x7fdd74009240 "CREATE INDEX CONCURRENTLY on prange1_2007_local(time);") at postgres.c:1105
#15 0x0000000102a54091 in PostgresMain (argc=1, argv=0x7fdd74003590, dbname=0x7fdd74003410 "postgres", username=0x7fdd740033e8 "onderkalaci") at postgres.c:4075
#16 0x000000010299c0fe in BackendRun (port=0x7fdd72e01aa0) at postmaster.c:4334
#17 0x000000010299b4f3 in BackendStartup (port=0x7fdd72e01aa0) at postmaster.c:4006
#18 0x000000010299a4f5 in ServerLoop () at postmaster.c:1743
#19 0x0000000102997d00 in PostmasterMain (argc=3, argv=0x7fdd72c03640) at postmaster.c:1351
#20 0x00000001028bb1df in main (argc=3, argv=0x7fdd72c03640) at main.c:228
(gdb) c
Continuing.
```

jasonmp85 commented on src/backend/distributed/commands/transmit.c:
```diff
@@ -54,7 +55,13 @@ RedirectCopyDataToRegularFile(const char *filename)
 		if (copyData->len > 0)
 		{
+#if (PG_VERSION_NUM >= 100000)
+			int appended = FileWrite(fileDesc, copyData->data, copyData->len,
```
Just here, or everywhere in this commit?

anarazel replied:
Well, everywhere where we're waiting on IO, rather than latches/locks.

jasonmp85 replied:
Done.

anarazel commented:
On 2017-06-12 11:38:51 -0700, Jason Petersen wrote:
> jasonmp85 commented on this pull request.
> 
> 
> 
> >  
> +#if (PG_VERSION_NUM >= 100000)
> +	CatalogTupleUpdate(pgDistColocation, &newHeapTuple->t_self, newHeapTuple);
> +#else
> +	simple_heap_update(pgDistColocation, &newHeapTuple->t_self, newHeapTuple);
>  	CatalogUpdateIndexes(pgDistColocation, newHeapTuple);
> 
> Done.

Why are you using macros instead of inline functions (just just plain
functions) here?  They have multiple evaluation hazards, worse error
messages, etc. pp?

jasonmp85 commented:
OK, pushed up commits addressing all of @anarazel's concerns. Still looking into @onderkalaci's crasher and error message concern.

anarazel commented on src/backend/distributed/utils/ruleutils_10.c:
```diff
@@ -0,0 +1,7745 @@
+/*-------------------------------------------------------------------------
+ *
+ * ruleutils.c
```
Version number in filename...

jasonmp85 replied:
I very much dislike having the filename inside the file. Mostly useless and annoying to maintain. Will fix.

jasonmp85 replied:
Oops, almost forgot this small change. Fixed.

anarazel commented:
On 2017-06-14 13:58:02 -0700, Jason Petersen wrote:
> jasonmp85 commented on this pull request.
> 
> 
> 
> > @@ -0,0 +1,7745 @@
> +/*-------------------------------------------------------------------------
> + *
> + * ruleutils.c
> 
> I very much dislike having the filename inside the file. Mostly useless and annoying to maintain. Will fix.

Yea, not a fan either.  But if we have, it should be accurate.

jasonmp85 commented:
After you finish each round of a review, can you add some sort of top-level comment letting me know you're done with a pass? I see comments appear but am not sure when they'll stop coming in.

anarazel commented:
> After you finish each round of a review, can you add some sort of top-level comment letting me know you're done with a pass? I see comments appear but am not sure when they'll stop coming in.

Ok, will done.  I'm done for now ;)

jasonmp85 commented:
> Ok, will do. I'm done for now

OK, so…? Please be more explicit both about when you finish adding comments and about whether (a) it's a :shipit: or (b) if it's not, what must be changed to ship.

I'm assuming it's the quicker-fix I implemented that @marcocitus suggested, but given that that discussion is buried in a thread halfway up the screen, it's always nice to have a summary comment to emphasize what is actually important to the progression of the review.

anarazel commented:
I think it's pretty clear that it's not yet ready to ship?  Clearly I don't think it's ok to ship-it without a proper fix for the memory-lifetime issue, there's still references to temporary branches ("pg10_hack"), the tests don't actually pass ("configure: error: Could not detect PostgreSQL version from pg_config.").    If the branch clearly isn't in a state that's mergeable, I don't quite see the point in detailing why exactly that's the case?

jasonmp85 commented:
> I don't quite see the point in detailing why exactly that's the case?

jasonmp85 commented:
> I don't quite see the point in detailing why exactly that's the case?

It's about review protocol; leaving a bunch of comments with no indication of completion or the relative importance of each one leaves the submitter sitting around wondering what to tackle next, or whether more comments are coming in, etc.

It sounds like the memory lifetime thing is it (the test problems you detail will go away when I merge the other branch into this one, and I'm similarly fixing the tools branch as well).

sumedhpathak commented:
@jasonmp85 is there an update on the crash? Whats the estimate on getting this fixed?

jasonmp85 commented:
Hm. I realized while fixing @onderkalaci's crash that it has nothing to do with PostgreSQL 10, `PARTITION` tables, or this patch. In fact, it's present in `6.2`. Opening another issue to track it, since we might want to backport.

onderkalaci commented:
> Hm. I realized while fixing @onderkalaci's crash that it has nothing to do with PostgreSQL 10, PARTITION tables, or this patch. In fact, it's present in 6.2. Opening another issue to track it, since we might want to backport.

Well, sorry for not elaborating this in detail before posting here. 

I've hit another issue on the branch, which I definitely wasn't hitting previously. I also reproduced this on a clean/separate ec2 instance to make sure this time not posting anything unrelated.

As I mentioned, I was using this branch for a while, and this was working for sure. The steps are as simple as below:

1. Download & install PG10 Beta
2. Checkout to this branch 
3. Create extension & add to preload libraries
4. Create a partitioned table

```SQL
CREATE TABLE prange2(id int, time date) PARTITION BY RANGE (time);
server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.

```

bt:
```

Program received signal EXC_BAD_ACCESS, Could not access memory.
Reason: KERN_INVALID_ADDRESS at address: 0x0000000000000010
EventTriggerAlterTableRelid (objectId=24576) at event_trigger.c:1752
1752		currentEventTriggerState->currentCommand->d.alterTable.objectId = objectId;
(gdb) bt
#0  EventTriggerAlterTableRelid (objectId=24576) at event_trigger.c:1752
#1  0x000000010cdbfe37 in AlterTableInternal (relid=24576, cmds=0x7f86498015b8, recurse=1 '\001') at tablecmds.c:3135
#2  0x000000010cdbd631 in DefineRelation (stmt=0x7f864a0021b0, relkind=112 'p', ownerId=10, typaddress=0x0, queryString=0x7f8649074640 "CREATE TABLE date_partitioned_table(id int, time date) PARTITION BY RANGE (time);") at tablecmds.c:856
#3  0x000000010d0254bd in ProcessUtilitySlow (pstate=0x7f864a002098, pstmt=0x7f8649075a80, queryString=0x7f8649074640 "CREATE TABLE date_partitioned_table(id int, time date) PARTITION BY RANGE (time);", context=PROCESS_UTILITY_TOPLEVEL, params=0x0, queryEnv=0x0, dest=0x7f8649075b78, completionTag=0x7fff530963d0 "") at utility.c:1008
#4  0x000000010d024da8 in standard_ProcessUtility (pstmt=0x7f8649075a80, queryString=0x7f8649074640 "CREATE TABLE date_partitioned_table(id int, time date) PARTITION BY RANGE (time);", context=PROCESS_UTILITY_TOPLEVEL, params=0x0, queryEnv=0x0, dest=0x7f8649075b78, completionTag=0x7fff530963d0 "") at utility.c:928
#5  0x000000010d91caa5 in multi_ProcessUtility (pstmt=0x7f8649075a80, queryString=0x7f8649074640 "CREATE TABLE date_partitioned_table(id int, time date) PARTITION BY RANGE (time);", context=PROCESS_UTILITY_TOPLEVEL, params=0x0, queryEnv=0x0, dest=0x7f8649075b78, completionTag=0x7fff530963d0 "") at executor/multi_utility.c:460
#6  0x000000010d023d8c in ProcessUtility (pstmt=0x7f8649075a80, queryString=0x7f8649074640 "CREATE TABLE date_partitioned_table(id int, time date) PARTITION BY RANGE (time);", context=PROCESS_UTILITY_TOPLEVEL, params=0x0, queryEnv=0x0, dest=0x7f8649075b78, completionTag=0x7fff530963d0 "") at utility.c:353
#7  0x000000010d023688 in PortalRunUtility (portal=0x7f86490c2440, pstmt=0x7f8649075a80, isTopLevel=1 '\001', setHoldSnapshot=0 '\0', dest=0x7f8649075b78, completionTag=0x7fff530963d0 "") at pquery.c:1179
#8  0x000000010d022896 in PortalRunMulti (portal=0x7f86490c2440, isTopLevel=1 '\001', setHoldSnapshot=0 '\0', dest=0x7f8649075b78, altdest=0x7f8649075b78, completionTag=0x7fff530963d0 "") at pquery.c:1325
#9  0x000000010d021ec4 in PortalRun (portal=0x7f86490c2440, count=9223372036854775807, isTopLevel=1 '\001', run_once=1 '\001', dest=0x7f8649075b78, altdest=0x7f8649075b78, completionTag=0x7fff530963d0 "") at pquery.c:800
#10 0x000000010d01d74c in exec_simple_query (query_string=0x7f8649074640 "CREATE TABLE date_partitioned_table(id int, time date) PARTITION BY RANGE (time);") at postgres.c:1105
#11 0x000000010d01c911 in PostgresMain (argc=1, argv=0x7f8649018220, dbname=0x7f8649018080 "postgres", username=0x7f8649018058 "onderkalaci") at postgres.c:4075
#12 0x000000010cf64bae in BackendRun (port=0x7f86487001f0) at postmaster.c:4334
#13 0x000000010cf63fa3 in BackendStartup (port=0x7f86487001f0) at postmaster.c:4006
#14 0x000000010cf62fa5 in ServerLoop () at postmaster.c:1743
#15 0x000000010cf607b0 in PostmasterMain (argc=3, argv=0x7f86484036a0) at postmaster.c:1351
#16 0x000000010ce83d8f in main (argc=3, argv=0x7f86484036a0) at main.c:228
```

When I drop Citus and remove from preload libraries, the crash goes away.

Similarly, when I run `check-vanilla`, the tests pass as @lithp mentioned [here](https://github.com/citusdata/citus/pull/1457#issuecomment-309375580). However, when I try an example from the Postgres regression tests, this still fails (which is interesting):

```SQL
CREATE TABLE partitioned (
        a int,
        b int,
        c text,
        d text
) PARTITION BY RANGE (a oid_ops, plusone(b), c collate "default", d collate "C");
2017-06-19 11:04:15.071 UTC [16261] LOG:  terminating any other active server processes
2017-06-19 11:04:15.071 UTC [17280] WARNING:  terminating connection because of crash of another server process
2017-06-19 11:04:15.071 UTC [17280] DETAIL:  The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
2017-06-19 11:04:15.071 UTC [17280] HINT:  In a moment you should be able to reconnect to the database and repeat your command.
2017-06-19 11:04:15.071 UTC [18260] FATAL:  the database system is in recovery mode
```

I've rebased this branch today, and it started crashing. I'm curious whether this is something that randomly happens or have we changed anything recently on the branch?

onderkalaci commented:
(Also, the issue is apparent on [this](https://travis-ci.org/citusdata/citus/jobs/244425421) Travis build as well)

jasonmp85 commented:
> The steps are as simple as below…

@onderkalaci I have yet to be able to reproduce this `PARTITION`-related crash on my Mac. Spinning up a Linux box to check out there, though I don't immediately see why we'd have a crash on Linux but not macOS (most bad access crashes like the above trigger just fine on macOS and leave a stack trace in Console.app).

onderkalaci commented:
> When I drop Citus and remove from preload libraries, the crash goes away.

It looks like the crash was happening due to the differences/incompatibilities between Citus and Postgres versions. Because after rebasing to this branch and using Postgres main branch (where I was using beta1 tag), the crash goes away.

I tested with MacOs, amazon instance and ubuntu.

(Sorry for the trouble.)

lithp commented:
Just ran pg's installcheck against this branch with `--load-extension=citus` and didn't find anything worrying

jasonmp85 commented:
@anarazel I believe I've addressed any and all threads from above. Let me know whether anything else is outstanding.

anarazel commented:
> @anarazel I believe I've addressed any and all threads from above. Let me know whether anything else is outstanding.

Are you planning to rebase/resplit the series?

jasonmp85 commented:
> Are you planning to rebase/resplit the series?

I don't entirely understand the question; it's been recently rebased, so it's up-to-date. What do I need to split?

For merge I assume I'll squash.

anarazel commented on .travis.yml:
```diff
@@ -16,18 +16,16 @@ env:
   - install_pg
   - install_custom_pg
-  # download and install HLL manually, as custom builds won't satisfy deps
-  - apt-get download "postgresql-${PGVERSION}-hll=2.10.1.citus-1" && sudo dpkg --force-confold --force-confdef --force-all -i *hll*.deb
```
Hm, what's the plan here? Or is that now solved in a different manner?

jasonmp85 replied:
We haven't started building packages for HyperLogLog yet. I suppose we have two approaches:

  * Block this review on creating a HyperLogLog package for PostgreSQL 10
    * Quick approach: just build a one-off package for Ubuntu Trusty/PostgreSQL 10 and throw it in packagecloud to satisfy this line
    * Right approach: update `citus_package` to build for PostgreSQL 10
  * Leave this as-is (there are test variants that pass when HyperLogLog is absent) and add back when PostgreSQL 10 is released
  * Expand this line into a script that gracefully continues when HyperLogLog can't be installed (preserves existing behavior for 9.5/9.6 and punts on PostgreSQL 10)

Pick one and I'll do it.

anarazel replied:
I think I can live with any approach but "leave this as-is", although I prefer not to do that !v10 approach.  Seems too likely that we'll drop the ball and this will bitrot.

anarazel commented on configure.in:
```diff
@@ -50,13 +50,13 @@ fi
 citusac_pg_config_version=$($PG_CONFIG --version 2>/dev/null)
 version_num=$(echo "$citusac_pg_config_version"|
-              $SED -e 's/^PostgreSQL \([[0-9]]*\)\.\([[0-9]]*\)\([[a-zA-Z0-9.]]*\)$/\1.\2/')
+              $SED -e 's/^PostgreSQL \([[0-9]]*\)\(\.[[0-9]]*\)\{0,1\}\(.*\)$/\1\2/')
```
Isn't this going to break the moment postgres 10 is released? Because then we'll have 10.0....

jasonmp85 replied:
I think this comment should probably be on the actual test a few lines down, since this regex should match `10.0` just fine. I'll add a `10.0` case to the test a few lines down to permit `10.0` once it's released (is that sufficient?).

anarazel replied:
No, that doesn't solve the issue, because then we'll have to adjust the expect lines for every following patch release (which are released every 2-3 months).  Remember, 10 dropped one number from the numbering schemes.  It used to be that 9.5 -> 9.6 were "major" (feature adding) releases, and 9.5.0 -> 9.5.1 were patch/minor releases (bugfix only).  There's now one digit less. I.e. 10->11 will be a major release, and 10.0 -> 10.1 -> 10.2 will be minor releases (bugfix only).

I don't think it makes sense to address that in the test block below.

anarazel commented on src/backend/distributed/commands/create_distributed_table.c:
```diff
@@ -349,6 +349,15 @@ ConvertToDistributedTable(Oid relationId, char *distributionColumnName,
+		ereport(ERROR, (errcode(ERRCODE_WRONG_OBJECT_TYPE),
+						errmsg("cannot distribute relation: %s", relationName),
+						errdetail("Distributing partition tables is unsupported.")));
+	}
```
This should have a test.  I think this is not sufficient, because it won't catch a partitioned table, just it's partition. Seems we need both for now.

jasonmp85 replied:
It's sufficient because partitioned tables are caught by the above check (see @onderkalaci's original comment about this); they're not regular or foreign tables, so the existing error is straightforward.

If you care, I can add a special error for PostgreSQL 10 which looks for that type, but at the moment we don't handle them specially for 9.5 or 9.6 and the above "is this regular or foreign?" check catches them just fine in 10.

jasonmp85 replied:
(However, agreed on the test, though I didn't want to add a test variant just for this one syntax thing @onderkalaci is likely to remove soon anyhow. That said, I'll just add an entirely new test file for PostgreSQL 10-specific syntax features).

jasonmp85 commented:
Checklist for myself:

  - [ ] Add `10.0` to the list of accepted/detected PostgreSQL versions
  - [ ] Build one-off Ubuntu Trusty PostgreSQL 10 HyperLogLog package (unless you want other solution)
  - [ ] Address sequence feedback I'd missed
  - [ ] Add tests for partition tables in new test file, with PostgreSQL 9.x/10 variants

Did I miss anything else?

onderkalaci commented:
>  Add tests for partition tables in new test file, with PostgreSQL 9.x/10 variants

I think you should skip adding more checks for partitioning. I already opened a PR (#1463) on top of this, which does that. It looks we're going to merge that PR right after this. (We're already exploring all edge cases in the scope of #1459)

jasonmp85 commented:
Minor diversion: #1402 broke this PR; fixing.

jasonmp85 commented:
OK, so I looked into supporting `GENERATED ... AS IDENTITY` but am passing on it for now. Instead I've modified `ConvertToDistributedTable` (the chokepoint for all distributed table creation) to reject tables which use this feature. I also found both places that used e.g. `sequenceIsOwned` and modified them to also find internal dependencies (though the first change should keep the second from ever being hit).

I've added a test for this.

jasonmp85 commented:
@anarazel fixed the version stuff, sequences stuff, and had to fix a few things for compatibility with latest `master` changes. Now there's [this bizarre failure](https://travis-ci.org/citusdata/citus/jobs/246073789), only on the Pull Request, and only on the PostgreSQL 10 source build. not sure what's causing it (I don't see recent PostgreSQL nor Citus changes that appear remotely related, but I could have missed something).

jasonmp85 commented:
On a hunch I'm clearing the Travis build cache directories to see whether that might fix anything.

jasonmp85 commented:
OK, I don't know why or how but that fixed the build. 🤷‍♂️

anarazel commented:
> OK, so I looked into supporting GENERATED ... AS IDENTITY but am passing on it for now. Instead I've modified ConvertToDistributedTable (the chokepoint for all distributed table creation) to reject tables which use this feature.

That's quite reasonable.

anarazel commented:
I still think it'd be better to have tests for the partitioned table rejection, but I guess @onderkalaci was against that...  I can live with that.
